### PR TITLE
Job Scheduler: Performance Improvements

### PR DESCRIPTION
**Pull request by hugoghx**
This PR contains work pertaining to improving database operation performance on the job scheduler:

1. Remove job_run db entries as soon as the jobs successfully complete;
2. Update the `RunJobs` query to not use the `job_jobs_to_run` view as its definition was causing database contention in the `job_run` table

### CODE REVIEW

File: internal/db/schema/migrations/oss/postgres/93/01_job_run_clean.up.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  -- remove 'completed' from the valid statuses.
+  -- updates 7/03_job.up.sql.
+  delete from job_run_status_enm where name = 'completed';
+
+  alter table job_run_status_enm
+    drop constraint only_predefined_job_status_allowed;
+
+  alter table job_run_status_enm
+    add constraint only_predefined_job_status_allowed
+      check(name in ('running', 'failed', 'interrupted'));
```
Comment (tmessi) [inline internal/db/schema/migrations/oss/postgres/93/01_job_run_clean.up.sql:34]: You can combine these into one `alter table` statement, ie:

```suggestion
  alter table job_run_status_enm
    drop constraint only_predefined_job_status_allowed,
    add  constraint only_predefined_job_status_allowed
      check(name in ('running', 'failed', 'interrupted'));
```

Comment (hugoghx) [reply internal/db/schema/migrations/oss/postgres/93/01_job_run_clean.up.sql]: Good call - Updated with https://github.com/hashicorp/boundary/pull/5204/commits/73ec9f709693a4bf6134102569286c8cfec1ae20

**Review by louisruch — CHANGES_REQUESTED** [review]
Looks good left a couple comments

**Comment by louisruch** [comment]
Mike merged a fairly extensive PR related to jobs so this is now also conflicting

File: internal/scheduler/job/run.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 	"google.golang.org/protobuf/proto"
 )
 
-// Run represents an instance of a job that is either actively running or has already completed.
+// Run represents an instance of a job that is either actively running or has failed in some way.'
```
[the remark this answers is not in the document]
Comment (hugoghx) [reply internal/scheduler/job/run.go]: https://github.com/hashicorp/boundary/pull/5204/commits/97722b56f417f9c3b40ae19d33a318a5746780fd

File: internal/scheduler/additional_verification_test.go [hunk 3]
```diff
@@ -619,7 +612,34 @@ func TestSchedulerRunNow(t *testing.T) {
 	close(jobCh)
 }
 
-func waitForRunStatus(t *testing.T, repo *job.Repository, runId, status string) *job.Run {
+func runCompleteFn(t *testing.T, sched *Scheduler, repo *job.Repository, runId, jobName string) func() {
```
[the remark this answers is not in the document]
Comment (hugoghx) [reply internal/scheduler/additional_verification_test.go]: Good point - Replaced name with `waitForRunComplete`: https://github.com/hashicorp/boundary/pull/5204/commits/74b11da856800a6abe72770e433b1cbb24ae74a6

File: internal/db/schema/migrations/oss/postgres/93/01_job_run_clean.up.sql [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+-- handles this by itself if the job is successful. It is possible that some
+-- entries are left in the table with this change (eg: Boundary is stopped after
+-- some jobs run but before the cleaner job runs).
+--
+-- These entries would forever be stored, so this migration cleans them to
+-- ensure no dangling rows are left behind.
+--
+-- It also updates the valid statues enum to reflect the ones in use.
+
+begin;
```
[the remark this answers is not in the document]
Comment (hugoghx) [reply internal/db/schema/migrations/oss/postgres/93/01_job_run_clean.up.sql:15]: Good catch! Deleted any `job_run_cleaner` entries from `job` and `job_run` tables: https://github.com/hashicorp/boundary/pull/5204/commits/de12b03536c743ba1695f2ddd9bdca2c3111fc5b

File: internal/scheduler/job/repository_run_test.go [hunk 5]
```diff
@@ -592,26 +559,16 @@ func TestRepository_CompleteRun(t *testing.T) {
...
 
-			// The previous run is ended before the next run is scheduled, therefore the previous
-			// run end time incremented by the nextRunIn duration, should be less than or equal to the
-			// NextScheduledRun time that is persisted in the repo.
-			nextRunAt := updatedJob.NextScheduledRun.AsTime()
-			previousRunEnd := got.EndTime.AsTime()
-			assert.Equal(nextRunAt.Round(time.Minute), previousRunEnd.Add(tt.nextRunIn).Round(time.Minute))
-
-			// Delete job run so it does not clash with future runs
-			_, err = repo.deleteRun(ctx, privateId)
```
[the remark this answers is not in the document]
Comment (hugoghx) [reply internal/scheduler/job/repository_run_test.go:613]: Yep! Added more validation: https://github.com/hashicorp/boundary/pull/5204/commits/9028a1e3c0e974db0aebbd0ff6b2e72b51c86b97

File: internal/scheduler/job/query.go [hunk 6]
```diff
@@ -74,17 +74,10 @@ const updateProgressQuery = `
...
-	  completed_count = ?,
-	  total_count     = ?,
-	  retries_count   = ?,
-	  status          = 'completed',
-	  end_time        = current_timestamp
+	delete from job_run
 	where
-	  private_id = ?
-	  and status = 'running'
+		private_id = ?
```
Comment (louisruch) [inline internal/scheduler/job/query.go:79]: We normally apply our sql style [guide](https://github.com/hashicorp/boundary/blob/main/sql-style-guide.md#formatting) to `query.go` files so thses should not be tabs

**Review by louisruch — COMMENTED** [review]
One style comment otherwise this looks great

File: internal/scheduler/job/query.go (hunk 6 above)
Comment (hugoghx) [reply internal/scheduler/job/query.go]: Addressed in https://github.com/hashicorp/boundary/pull/5204/commits/27c432a74853cad9c69e4d4f44fb87d9cadb9ec7

**Review by louisruch — APPROVED** [review]
LGTM, @tmessi should be back tomorrow so might be good to wait for him to 👍 since he did have some comments

**Review by tmessi — COMMENTED** [review]
I think this looks good, can you move it out of draft so the schema-diff workflow runs? It would be nice to review the schema diff just to be sure.

**Comment by hugoghx** [comment]
> I think this looks good, can you move it out of draft so the schema-diff workflow runs? It would be nice to review the schema diff just to be sure.

Yep, done!

I will also add the complete query change to this PR, likely later today

**Comment by hugoghx** [comment]
@tmessi @louisruch Pushed up the changes to the job scheduling logic: https://github.com/hashicorp/boundary/pull/5204/commits/f227020b670cef07ece40d121bea826f8d17e0be

**Review by tmessi — APPROVED** [review]
Nice work!

**Review by louisruch — APPROVED** [review]
:shipit:

**Pull request merged**